src/main/java/com/google/api/codegen/transformer/csharp/CSharpSurfaceNamer.java

```diff
@@ -0,0 +1,174 @@
+  }
+
+  @Override
+  public String getStaticLangAsyncReturnTypeName(Method method, MethodConfig methodConfig) {
+    if (ServiceMessages.s_isEmptyType(method.getOutputType())) {
+      return "Task";
+    }
+    return "Task<" + getModelTypeFormatter().getFullNameFor(method.getOutputType()) + ">";
+  }
+
+  @Override
+  public String getApiWrapperClassName(Interface interfaze) {
+    return className(Name.upperCamel(interfaze.getSimpleName(), "Client"));
+  }
+
+  @Override
+  public String getCallableName(Method method) {
+    // TODO: Use the 'privateFieldName' method when it's available (from Go MVVM PR)
+    return "_" + varName(Name.upperCamel("Call", method.getSimpleName()));
+  }
+
+  @Override
+  public String getPathTemplateName(CollectionConfig collectionConfig) {
+    return inittedConstantName(Name.from(collectionConfig.getEntityName(), "template"));
+  }
+
+  @Override
+  public String getFieldGetFunctionName(TypeRef type, Name identifier) {
+    return methodName(identifier);
+  }
+
+  @Override
+  public String getAndSavePagedResponseTypeName(
+      ModelTypeTable typeTable, TypeRef... parameterizedTypes) {
+    String[] typeList = new String[parameterizedTypes.length];
+    for (int i = 0; i < typeList.length; i++) {
+      typeList[i] = typeTable.getFullNameForElementType(parameterizedTypes[i]);
+    }
+    return typeTable.getAndSaveNicknameForContainer("Google.Api.Gax.PagedEnumerable", typeList);
+  }
+
+  @Override
+  public String getGrpcContainerTypeName(Interface service) {
+    return className(Name.upperCamel(service.getSimpleName()));
+  }
+
+  @Override
+  public String getGrpcServiceClassName(Interface service) {
+    return className(Name.upperCamel(service.getSimpleName()))
+        + "."
+        + className(Name.upperCamel(service.getSimpleName(), "Client"));
+  }
+
+  @Override
+  public String getApiWrapperClassImplName(Interface interfaze) {
+    return className(Name.upperCamel(interfaze.getSimpleName(), "ClientImpl"));
+  }
+
+  @Override
+  public String getAsyncApiMethodName(Method method) {
+    return getApiMethodName(method) + "Async";
+  }
+
+  @Override
+  public String getPageStreamingDescriptorConstName(Method method) {
+    return inittedConstantName(Name.upperCamel(method.getSimpleName()));
+  }
+
+  @Override
+  public String getParamName(String var) {
+    return varName(Name.from(var).join("id"));
+  }
+
+  @Override
+  public String retryFilterMethodName(String key) {
+    return methodName(Name.from(key).join("retry").join("filter"));
+  }
+
+  /** The method name of the retry backoff for the given key */
+  @Override
+  public String retryBackoffMethodName(String key) {
+    return methodName(Name.from("get").join(key).join("retry").join("backoff"));
+  }
+
+  /** The method name of the timeout backoff for the given key */
+  @Override
+  public String timeoutBackoffMethodName(String key) {
+    return methodName(Name.from("get").join(key).join("timeout").join("backoff"));
+  }
```

Comment:
My preference is for functions that purely provide names to have a default implementation in SurfaceNamer instead of just using the not-implemented string. That way, if someone is generating for a similar scenario, they could re-use the names. I have tried to use the not-implemented string when there is not a sensible default. Your default names for the methods above seem sensible to me for SurfaceNamer.